doc: add auto remount volume page (#874)
longhorn/website: content/docs/1.5.4/high-availability/recover-volume.md

---
  title: Recover Volume
  weight: 1
---

> ## Automatically deletes the workload pod

Author: Fixed

> When the following situations occur, Longhorn automatically attempts to delete workload pods that are managed by a controller (for example, Deployment, StatefulSet, or DaemonSet). After deletion, the controller restarts the workload pod and Kubernetes handles volume reattachment and remounting.

Author: Added, thanks

Longhorn does not delete pods without a controller because such pods cannot be restarted after deletion. To recover volumes that are unexpectedly detached, you must manually delete and restart the pods without a controller.

## Automatically remount the volume

The state of a volume can change to read-only when IO errors occur. IO errors can be caused by a variety of issues, including the following:
- Network disconnection: Interrupted connection between the engine and replicas.
- High disk latency: Significant delay in the transfer of data between a replica and the corresponding disk.

Longhorn checks the volume global mount point state every 10 sec. When the filesystem of the volume changes to read-only, Longhorn automatically attempts to remount the global mount point on the host to change the state back to read-write. Upon successful remounting, the workload pods continue functioning without disruption.

> ## Conclusion

Reviewer: More suggestions and info can be provided in this section. 
IIUC, IO errors may lead to engine crash then can be handled by `Automatically deletes the workload pod`. In this case, does users need to enable the setting?

Author: `Automatically remount the volume` this feature doesn't have setting to enable/disable it.
It just blindly tries to remount the `globalmount` when detecting `ro`

For engine crash case,
1. Longhorn will detach and reattach the the volume (no setting, default behavior)
2. The volume will become healthy again
3. However, the K8s won't perform remount in this case because the workload pod is still running 
4. User won't be able to perform IO and will get `read-only filesystem` error and the filesystem become `read-only`

In this case, the `Automatically remount the volume`  won't help, 
because when it tries to remount the filesystem it will get 
```
/host/var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/fe727eb53a0f2a72f000ffa85e652a92674a30d43befce23d0468712717b0de1/globalmount: 
cannot remount /dev/longhorn/pvc-e83db2c5-569e-46e0-add9-15db7debdab2 read-write, is write-protected.
```

Thus, for **engine crash case**, user still needs to enable `Automatically deletes the workload pod` to request k8s to remount the workload.

cc @derekbit

Author: I have added the case study to the  `Automatically remount the volume` section

---

>   title: Recover Volume

Reviewer 2: ```suggestion
  title: Volume Recovery
```

  weight: 1
---

> This section is aimed to inform users of what might leads to volume failure and how Longhorn recover the volume.

Reviewer 2: ```suggestion
Longhorn provides two mechanisms for maintaining volume functionality in a variety of situations. 
```

3. An error occurred on a Share Manager pod that uses an RWX volume.

If you want to prevent Longhorn from automatically deleting workload pods, disable the setting [*Automatically Delete Workload Pod when The Volume Is Detached Unexpectedly*](../../references/settings#automatically-delete-workload-pod-when-the-volume-is-detached-unexpectedly) on the Longhorn UI.

Longhorn does not delete pods without a controller because such pods cannot be restarted after deletion. To recover volumes that are unexpectedly detached, you must manually delete and restart the pods without a controller.

## Automatically remount the volume

This recovery mechanism does not have setting to control it.

The state of a volume can change to read-only when IO errors occur. IO errors can be caused by a variety of issues, including the following:
- Network disconnection: Interrupted connection between the engine and replicas.
- High disk latency: Significant delay in the transfer of data between a replica and the corresponding disk.

> Longhorn checks the volume global mount point state every 10 sec. When the filesystem of the volume changes to read-only, Longhorn updates the condition to the engine of the volume for users to check. Longhorn then automatically attempts to remount the global mount point on the host to change the state back to read-write. Upon successful remounting, the workload pods continue functioning without disruption.

Reviewer 2: ```suggestion
Longhorn checks the state of the volume's global mount point every 10 seconds. When the volume's filesystem changes to read-only, Longhorn updates the condition to the volume's data engine. Longhorn then automatically attempts to remount the global mount point on the host to change the state back to read-write. Upon successful remounting, the workload pods continue functioning without disruption.
```

> Noted, this mechanism might not work for some cases. For example, when the engine of a volume crashes, Longhorn detaches and attaches the volume automatically. The filesystem will become read-only after reattaching and can not be remounted back to read-write. In this case, users still need to rely on [Automatically deletes the workload pod](#automatically-deletes-the-workload-pod) to request Kubernetes to remount the volume by recreating the pod.

Reviewer 2: ```suggestion
> **Note:**
> This mechanism might not work in some situations. For example, when the volume's data engine crashes, Longhorn automatically detaches and reattaches the volume. The filesystem changes to read-only after reattaching, and you cannot change it back to read-write. In this case, you can only rely on the [Automatic Workload Pod Deletion](#automatic-workload-pod-deletion) mechanism, which enables volume remounting after the workload pod is recreated.
```

> ## Conclusion

Reviewer 2: ```suggestion
## Summary
```

> These two mechanisms co-exists in Longhorn to handle different situations and maintain volume functionality. 

Reviewer 2: This is not necessary. Please remove it.

> [Automatically deletes the workload pod](#automatically-deletes-the-workload-pod) is triggered when unexpected failures happen. By deleting the workload pod, the controller restarts the workload pod and Kubernetes handles volume reattachment and remounting. This introduce some interruption to the workload. One can disable the function by disabling the setting 
> [*Automatically Delete Workload Pod when The Volume Is Detached Unexpectedly*](../../references/settings#automatically-delete-workload-pod-when-the-volume-is-detached-unexpectedly) on the Longhorn UI.

Reviewer 2: ```suggestion
[Automatic Workload Pod Deletion](#automatic-workload-pod-deletion) is triggered when unexpected failures happen. The controller deletes and then restarts the workload pod, and Kubernetes handles volume reattachment and remounting. The process may cause interruptions to the workload. If you want to prevent Longhorn from automatically deleting workload pods, disable the setting [*Automatically Delete Workload Pod when The Volume Is Detached Unexpectedly*](../../references/settings#automatically-delete-workload-pod-when-the-volume-is-detached-unexpectedly) on the Longhorn UI.
```

> [Automatically remount the volume](#automatically-remount-the-volume) is triggered when Longhorn finds the filesystem of the volume changes to read-only. Longhorn then remounts the global mount point on the host to change the state back to read-write.

Reviewer 2: ```suggestion
[Automatic Volume Remounting](#automatic-volume-remounting) is triggered when the volume's filesystem changes to read-only. Longhorn remounts the global mount point on the host to change the state back to read-write.
```
